examples/server/server.cpp

```diff
@@ -3159,6 +3275,112 @@ int main(int argc, char ** argv) {
         res.status = 200; // HTTP OK
     };
 
+    const auto handle_slots_save = [&ctx_server, &res_error, &sparams](const httplib::Request & req, httplib::Response & res, int id_slot) {
+        json request_data = json::parse(req.body);
+        std::string filename = request_data["filename"];
+        if (filename.find('/') != std::string::npos || filename.find('\\') != std::string::npos || filename.find("..") != std::string::npos) {
```

Comment:
This won't take into account all possible edge cases, for example: [using unicode](https://github.com/swisskyrepo/PayloadsAllTheThings/blob/master/Directory%20Traversal/README.md)

I couldn't find any lightweight implementation of file name sanitizer function on cpp, but maybe we can base on [this js code](https://github.com/parshap/node-sanitize-filename/blob/master/index.js) and make our own `sanitize_file_name(std::string filename)`, place it in `utils.hpp`

Reply:
Yes, let's simplify this - maybe even hardcode this if it becomes too complicated. We don't want to be extremely flexible in this case, but we do want to avoid and complications when dealing with arbitrary user input

Reply:
Rather than sanitize, just completely reject any filename with overlong UTF-8 encodings. The code is already blocking any ".." rather than "../" style combinations, so based on those documents, (in addition to rejecting any overlong UTF-8) just the uff0e, u2215, and u2216 unicode codepoints still need to be blocked.

Reply:
Yeah the idea seems ok to me. I mean we can borrow the sanitize code to block certain patterns here. At worst, we can simply allow all ASCII alphanumeric characters (whitelisting instead of blacklisting)

Reply:
Added validation functions. Here's the test cases, not sure where to add them in:

```
    // files
    test(validate_file_name("file.bin") == true);
    test(validate_file_name(u8"日本式ファイルの芸術.bin") == true);

    // paths
    test(validate_file_name("../bad.bin") == false);
    test(validate_file_name("..\\bad.bin") == false);
    test(validate_file_name("also/..\\bad.bin") == false);
    test(validate_file_name("also/bad.bin") == false);

    // unicode equivalents
    test(validate_file_name("unicode" "\xE2\x88\x95" "bad.bin") == false); // u2215
    test(validate_file_name("unicode" "\xE2\x88\x96" "bad.bin") == false); // u2216
    test(validate_file_name("unicode" "\xEF\xBC\x8E" "bad.bin") == false); // uFF0E

    // overlong encoding
    test(validate_file_name("overlong" "\xC0\x2E" "bad.bin") == false);
    test(validate_file_name("overlong" "\xE0\x40\xAE" "bad.bin") == false);
    test(validate_file_name("overlong" "\xC0\xAE" "bad.bin") == false);
    test(validate_file_name("overlong" "\xC0\xAF" "bad.bin") == false);
    test(validate_file_name("overlong" "\xE0\x80\xAF" "bad.bin") == false);
    test(validate_file_name("overlong" "\xC0\x2F" "bad.bin") == false);
    test(validate_file_name("overlong" "\xC0\x5C" "bad.bin") == false);
    test(validate_file_name("overlong" "\xC0\x80\x5C" "bad.bin") == false);
```

Reply:
Also need to reject leading and trailing 0x20 whitespace and trailing dots, see https://learn.microsoft.com/en-us/troubleshoot/windows-client/shell-experience/file-folder-name-whitespace-characters.